modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C

```diff
@@ -0,0 +1,1229 @@
+  const Real y2 = p2.second;
+  const Real x3 = p3.first;
+  const Real y3 = p3.second;
+  const Real x4 = p4.first;
+  const Real y4 = p4.second;
+
+  Real x = -((x1 - x2) * (y3 * x4 - x3 * y4) - (x3 - x4) * (y1 * x2 - x1 * y2)) /
+           ((y1 - y2) * (x3 - x4) - (y3 - y4) * (x1 - x2));
+  Real y = -((y1 - y2) * (y3 * x4 - x3 * y4) - (y3 - y4) * (y1 * x2 - x1 * y2)) /
+           ((y1 - y2) * (x3 - x4) - (y3 - y4) * (x1 - x2));
+
+  return std::make_pair(x, y);
+}
+
+// 'azimuthal_angles_collector' collects sorted azimuthal angles of the external boundary.
+std::vector<Real>
+PolygonMeshGeneratorBase::azimuthal_angles_collector(const std::unique_ptr<ReplicatedMesh> mesh,
+                                                     const Real lower_azi,
+                                                     const Real upper_azi,
+                                                     const unsigned int return_type,
+                                                     const bool calculate_origin,
+                                                     const Real input_origin_x,
+                                                     const Real input_origin_y,
+                                                     const Real tol)
+{
+  std::vector<Real> azimuthal_output;
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> side_list =
+      mesh->get_boundary_info().build_side_list();
+  mesh->get_boundary_info().build_node_list_from_side_list();
+  std::vector<std::tuple<dof_id_type, boundary_id_type>> node_list =
+      mesh->get_boundary_info().build_node_list();
+
+  std::vector<Real> bd_x_list;
+  std::vector<Real> bd_y_list;
+  Real origin_x = 0.0;
+  Real origin_y = 0.0;
+  Real vol_tmp = 0.0;
+  Real tmp_azi;
+  const Real mid_azi = lower_azi <= upper_azi ? (lower_azi + upper_azi) / 2.0
+                                              : (lower_azi + upper_azi + 360.0) / 2.0;
+  for (unsigned int i = 0; i < node_list.size(); ++i)
+  {
+    if (std::get<1>(node_list[i]) == OUTER_SIDESET_ID)
+    {
+      bd_x_list.push_back((*mesh->node_ptr(std::get<0>(node_list[i])))(0));
+      bd_y_list.push_back((*mesh->node_ptr(std::get<0>(node_list[i])))(1));
+    }
+  }
+
+  if (calculate_origin)
+  {
+    // Iterate through elements to calculate the center of mass of the mesh, which is used as the
+    // origin.
+    for (libMesh::ReplicatedMesh::element_iterator elem_it = mesh->elements_begin();
+         elem_it != mesh->elements_end();
+         elem_it++)
+    {
+      const unsigned int n_elem_nodes = (*elem_it)->n_nodes();
+      if (n_elem_nodes == 3)
+      {
+        origin_x += ((*elem_it)->centroid())(0) * ((*elem_it)->volume());
+        origin_y += ((*elem_it)->centroid())(1) * ((*elem_it)->volume());
+      }
+      else // n_elem_nodes == 4
+      {
+        // LibMesh "centroid" is just average of all nodes, which is not the real centroid of a
+        // quadrilateral. Use algorithm described in
```

Comment:
i think we have the real centroid in libmesh implemented too 
@roystgnr

Reply:
maybe not at the time? This is a recent libmesh update

Reply:
Right; IIRC `Elem::true_centroid()` *just* went in to the libMesh update that got merged a week and a half ago.  `Elem::centroid()` is now deprecated (because we can't tell who really wanted a centroid and who really wanted the old behavior...) and users are expected to switch to either `true_centroid()` or `vertex_average()` depending on whether their algorithm cares enough about correctness to sometimes sacrifice speed or whether it's happy with just getting a single point *near* an element's center.